replaced the gravity compensator with inverse dynamics

siyuanfeng-tri opened the pull request:
0. replaced gravity compensator in systems/controllers with inverse dynamics (wrappers around RBT.inverseDynamics). 
1. replaced the PidWithGravityCompensator with InverseDynamicsController. The latter computes 
trq = invdyn(q, v, vd_d), where vd = kp * (q* - q) + kd * (v* - v) + ki * int(q* - q) + vd*. 
The PID gains are now on desired acceleration as opposed to directly on torque.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/robotlocomotion/drake/5339)
<!-- Reviewable:end -->

siyuanfeng-tri commented:
+@naveenoid for feature review.
cc @patmarion 

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5339#-:-Ke0oCpRJs2naOZ2YonS:bw6gpr7)*
<!-- Sent from Reviewable.io -->

naveenoid commented:
Thanks for this! One pass completed, PTAL.

---

Reviewed 20 of 22 files at r1.
Review status: 20 of 22 files reviewed at latest revision, 8 unresolved discussions.

---

*[drake/examples/kuka_iiwa_arm/BUILD, line 82 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5339#-Ke3CxYjgrrUfgZR6qXX:-Ke3CxYjgrrUfgZR6qXY:b-cd2kwi) ([raw file](https://github.com/robotlocomotion/drake/blob/a840692ca780d0757f7bc5f4fff36d924fdbadc6/drake/examples/kuka_iiwa_arm/BUILD#L82)):*
> ```Python
>         "//drake/systems/analysis:simulator",
>         "//drake/systems/controllers:pid_controller",
>         "//drake/systems/controllers:inverse_dynamics_controller",
> ```

nit: alphabetical order.

---

*[drake/examples/kuka_iiwa_arm/iiwa_common.cc, line 171 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5339#-Ke3CZOn5R6dK3wLFMoY:-Ke3CZOn5R6dK3wLFMoZ:bjhetyh) ([raw file](https://github.com/robotlocomotion/drake/blob/a840692ca780d0757f7bc5f4fff36d924fdbadc6/drake/examples/kuka_iiwa_arm/iiwa_common.cc#L171)):*
> ```cpp
>                                     Eigen::VectorXd* Kd) {
>   Kp->resize(7);
>   *Kp << 100, 100, 100, 100, 100, 100, 100;
> ```

Do we still require such large gains? Also, why the large gains on the last few (low inertia) Dofs?

---

*[drake/examples/kuka_iiwa_arm/iiwa_wsg_simulation.cc, line 26 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5339#-Ke3CphGHlr-6BqESY8o:-Ke3CphGHlr-6BqESY8p:b-cd2kwi) ([raw file](https://github.com/robotlocomotion/drake/blob/a840692ca780d0757f7bc5f4fff36d924fdbadc6/drake/examples/kuka_iiwa_arm/iiwa_wsg_simulation.cc#L26)):*
> ```cpp
> #include "drake/systems/analysis/simulator.h"
> #include "drake/systems/controllers/pid_controller.h"
> #include "drake/systems/controllers/inverse_dynamics_controller.h"
> ```

nit: alphabetical order.

---

*[drake/examples/kuka_iiwa_arm/iiwa_wsg_simulation.cc, line 34 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5339#-Ke3DnyAXJ-7eywJYpOr:-Ke3DnyAXJ-7eywJYpOs:b-k8krls) ([raw file](https://github.com/robotlocomotion/drake/blob/a840692ca780d0757f7bc5f4fff36d924fdbadc6/drake/examples/kuka_iiwa_arm/iiwa_wsg_simulation.cc#L34)):*
> ```cpp
> #include "drake/systems/primitives/constant_vector_source.h"
> #include "drake/systems/primitives/matrix_gain.h"
> ```

BTW: Not a comment on your PR but I would ideally like to see this builder move to a separate header (FactoryBuilder type). Perhaps you can leave a TODO here if you agree? I can handle it.

---

*[drake/examples/kuka_iiwa_arm/iiwa_wsg_simulation.cc, line 168 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5339#-Ke3D7CRH6fhSr8HLQDl:-Ke3D7CRH6fhSr8HLQDm:b-hta1fa) ([raw file](https://github.com/robotlocomotion/drake/blob/a840692ca780d0757f7bc5f4fff36d924fdbadc6/drake/examples/kuka_iiwa_arm/iiwa_wsg_simulation.cc#L168)):*
> ```cpp
>     // estimation of the model parameters such as friction in the
>     // joints.
>     Eigen::VectorXd iiwa_kp = Eigen::VectorXd::Zero(7);
> ```

Can we replace the repetition in this segment with the method in iiwa_common? Or in case you want to use a integral component to compensate for the gripper, maybe that can be moved to iiwa_common as well? Would help maintaining the same tuning across demos.

---

*[drake/examples/kuka_iiwa_arm/iiwa_wsg_simulation.cc, line 180 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5339#-Ke3EBAbC7p8TxIZZZkI:-Ke3EBAbC7p8TxIZZZkJ:b4cs4t0) ([raw file](https://github.com/robotlocomotion/drake/blob/a840692ca780d0757f7bc5f4fff36d924fdbadc6/drake/examples/kuka_iiwa_arm/iiwa_wsg_simulation.cc#L180)):*
> ```cpp
>         builder.template AddSystem<systems::InverseDynamicsController<T>>(
>             iiwa_info.model_path, iiwa_info.world_offset, iiwa_kp, iiwa_ki,
>             iiwa_kd, false /* with feedforward acceleration */);
> ```

BTW: Just for clarity, a comment explaining why we choose to not use feedforward accelerations in this demo would be useful. 

---

*[drake/examples/kuka_iiwa_arm/controlled_kuka/kuka_demo_plant_builder.h, line 79 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5339#-Ke3EZJRPaFQ4OMVi9Us:-Ke3EZJRPaFQ4OMVi9Ut:b-6ayw0j) ([raw file](https://github.com/robotlocomotion/drake/blob/a840692ca780d0757f7bc5f4fff36d924fdbadc6/drake/examples/kuka_iiwa_arm/controlled_kuka/kuka_demo_plant_builder.h#L79)):*
> ```objc
>     // have any logic to calculate the desired target velocity (yet!)
>     // set the D term (to stabilize the arm near the commanded
>     // position) and feed a desired velocity vector of zero.
> ```

BTW: A comment of this type on feedforward accelerations above the IDController addsystem would be useful.

---

*[drake/systems/controllers/test/inverse_dynamics_test.cc, line 165 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5339#-Ke3H0wXdil0OwBZ3Ysc:-Ke3H0wXdil0OwBZ3Ysd:b-45sajy) ([raw file](https://github.com/robotlocomotion/drake/blob/a840692ca780d0757f7bc5f4fff36d924fdbadc6/drake/systems/controllers/test/inverse_dynamics_test.cc#L165)):*
> ```cpp
>       tree.get());
>   EXPECT_DEATH(Init(std::move(tree), true), ".*");
> }
> ```

BTW: Just for completeness, dont we need a death test for the PID dimensions?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5339#-:-Ke3HAWr3xwPIQy9eywb:b-9xeifl)*
<!-- Sent from Reviewable.io -->

siyuanfeng-tri commented:
Review status: 16 of 22 files reviewed at latest revision, 8 unresolved discussions.

---

*[drake/examples/kuka_iiwa_arm/BUILD, line 82 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5339#-Ke3CxYjgrrUfgZR6qXX:-Ke48AS_hNWH_sL0ak09:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/a840692ca780d0757f7bc5f4fff36d924fdbadc6/drake/examples/kuka_iiwa_arm/BUILD#L82)):*
<details><summary><i>Previously, naveenoid (Naveen Kuppuswamy) wrote…</i></summary><blockquote>

nit: alphabetical order.
</blockquote></details>

Done.

---

*[drake/examples/kuka_iiwa_arm/iiwa_common.cc, line 171 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5339#-Ke3CZOn5R6dK3wLFMoY:-Ke48CUu7QzvTyK0h5Fp:bg8o0jm) ([raw file](https://github.com/robotlocomotion/drake/blob/a840692ca780d0757f7bc5f4fff36d924fdbadc6/drake/examples/kuka_iiwa_arm/iiwa_common.cc#L171)):*
<details><summary><i>Previously, naveenoid (Naveen Kuppuswamy) wrote…</i></summary><blockquote>

Do we still require such large gains? Also, why the large gains on the last few (low inertia) Dofs?
</blockquote></details>

these are gains on accelerations now, the inertia stuff is taken case of by inverse dynamics. 
these are pretty arbitrarily set without any thinking. 

---

*[drake/examples/kuka_iiwa_arm/iiwa_wsg_simulation.cc, line 26 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5339#-Ke3CphGHlr-6BqESY8o:-Ke48vjAk4G5RUOfikQz:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/a840692ca780d0757f7bc5f4fff36d924fdbadc6/drake/examples/kuka_iiwa_arm/iiwa_wsg_simulation.cc#L26)):*
<details><summary><i>Previously, naveenoid (Naveen Kuppuswamy) wrote…</i></summary><blockquote>

nit: alphabetical order.
</blockquote></details>

Done.

---

*[drake/examples/kuka_iiwa_arm/iiwa_wsg_simulation.cc, line 34 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5339#-Ke3DnyAXJ-7eywJYpOr:-Ke49E7tpQLNKT49yQIE:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/a840692ca780d0757f7bc5f4fff36d924fdbadc6/drake/examples/kuka_iiwa_arm/iiwa_wsg_simulation.cc#L34)):*
<details><summary><i>Previously, naveenoid (Naveen Kuppuswamy) wrote…</i></summary><blockquote>

BTW: Not a comment on your PR but I would ideally like to see this builder move to a separate header (FactoryBuilder type). Perhaps you can leave a TODO here if you agree? I can handle it.
</blockquote></details>

Done.

---

*[drake/examples/kuka_iiwa_arm/iiwa_wsg_simulation.cc, line 168 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5339#-Ke3D7CRH6fhSr8HLQDl:-Ke4A-oZvJDykaPqvoH0:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/a840692ca780d0757f7bc5f4fff36d924fdbadc6/drake/examples/kuka_iiwa_arm/iiwa_wsg_simulation.cc#L168)):*
<details><summary><i>Previously, naveenoid (Naveen Kuppuswamy) wrote…</i></summary><blockquote>

Can we replace the repetition in this segment with the method in iiwa_common? Or in case you want to use a integral component to compensate for the gripper, maybe that can be moved to iiwa_common as well? Would help maintaining the same tuning across demos.
</blockquote></details>

Done.

---

*[drake/examples/kuka_iiwa_arm/iiwa_wsg_simulation.cc, line 180 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5339#-Ke3EBAbC7p8TxIZZZkI:-Ke4Aa1UUJZMDUtb3A8V:b667gwl) ([raw file](https://github.com/robotlocomotion/drake/blob/a840692ca780d0757f7bc5f4fff36d924fdbadc6/drake/examples/kuka_iiwa_arm/iiwa_wsg_simulation.cc#L180)):*
<details><summary><i>Previously, naveenoid (Naveen Kuppuswamy) wrote…</i></summary><blockquote>

BTW: Just for clarity, a comment explaining why we choose to not use feedforward accelerations in this demo would be useful. 
</blockquote></details>

we are exposing the ff vd, it's just set to zero currently, because kuka_plan_runner doesn't send out vd yet. will be an easy change later on. 

---

*[drake/examples/kuka_iiwa_arm/controlled_kuka/kuka_demo_plant_builder.h, line 79 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5339#-Ke3EZJRPaFQ4OMVi9Us:-Ke4BUPLbq9HDi0KnVQO:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/a840692ca780d0757f7bc5f4fff36d924fdbadc6/drake/examples/kuka_iiwa_arm/controlled_kuka/kuka_demo_plant_builder.h#L79)):*
<details><summary><i>Previously, naveenoid (Naveen Kuppuswamy) wrote…</i></summary><blockquote>

BTW: A comment of this type on feedforward accelerations above the IDController addsystem would be useful.
</blockquote></details>

Done.

---

*[drake/systems/controllers/test/inverse_dynamics_test.cc, line 165 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5339#-Ke3H0wXdil0OwBZ3Ysc:-Ke4BeaQv4Zb1kdUjIdM:b-ugskj4) ([raw file](https://github.com/robotlocomotion/drake/blob/a840692ca780d0757f7bc5f4fff36d924fdbadc6/drake/systems/controllers/test/inverse_dynamics_test.cc#L165)):*
<details><summary><i>Previously, naveenoid (Naveen Kuppuswamy) wrote…</i></summary><blockquote>

BTW: Just for completeness, dont we need a death test for the PID dimensions?
</blockquote></details>

for id, i think it's ok if q doesn't match v dimension wise, as long as it's fully actuated. i don't think we have a rocket engine powered block model? do we? 
the pid stuff should be in some other test.  

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5339)*
<!-- Sent from Reviewable.io -->

naveenoid commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 16 of 22 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5339#-:-Ke4KQeJItYVLBzIQwP8:bnfp4nl)*
<!-- Sent from Reviewable.io -->

siyuanfeng-tri commented:
+@sherm1 for platform review. thanks!

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5339#-:-Ke4Ku86bof3onIptGNi:bmp69bs)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
@drake-jenkins-bot retest this please

sherm1 commented:
Platform review pass done, PTAL. Questions:
- Why is it a good idea to get rid of the gravity compensator? It doesn't seem obvious to me that someone looking for gravity compensation will think to look at InverseDynamics. If it is because gravity compensation can be implemented using inverse dynamics, that seems like an implementation detail rather than something that a user would care about. In that case you could re-implement GravityCompensator using InverseDynamics under the covers.
- I couldn't find unit tests for InverseDynamicsController.
- The class documentation for InverseDynamics and InverseDynamicsController needs to be expanded, see below.

---

Reviewed 14 of 22 files at r1, 11 of 11 files at r2.
Review status: all files reviewed at latest revision, 12 unresolved discussions.

---

*[drake/examples/kuka_iiwa_arm/iiwa_wsg_simulation.cc, line 164 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5339#-Ke5xg7FUgMHTK7Nw5VK:-Ke5xg7FUgMHTK7Nw5VL:b-umv4xj) ([raw file](https://github.com/robotlocomotion/drake/blob/da4a09773e96859b322b90713c5e0790b3ce2ba9/drake/examples/kuka_iiwa_arm/iiwa_wsg_simulation.cc#L164)):*
> ```cpp
>     VectorX<double> iiwa_kp, iiwa_kd, iiwa_ki;
>     SetPositionControlledIiwaGains(&iiwa_kp, &iiwa_ki, &iiwa_kd);
>     // Have a bit of inertia gains to deal with the added mass from the grasped
> ```

BTW, "inertia gains"? Did you mean "integral gains"?

---

*[drake/examples/kuka_iiwa_arm/controlled_kuka/kuka_demo_plant_builder.h, line 74 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5339#-Ke5sSmK7FqtCyF2FnEK:-Ke5sSmK7FqtCyF2FnEL:bo5d60m) ([raw file](https://github.com/robotlocomotion/drake/blob/da4a09773e96859b322b90713c5e0790b3ce2ba9/drake/examples/kuka_iiwa_arm/controlled_kuka/kuka_demo_plant_builder.h#L74)):*
> ```objc
> 
>     // TODO(siyuan): should be able to easily compute derivatives once
>     // #5215 is in.
> ```

BTW, this comment seems to lack context now. Previously there was a discussion about the desirability of derivatives and why they aren't used here but now that is gone. Please explain (resurrect the previous comment?).

---

*[drake/systems/controllers/inverse_dynamics.cc, line 12 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5339#-Ke5trSzycDa_o2zxlub:-Ke5trSzycDa_o2zxluc:b-25x2b2) ([raw file](https://github.com/robotlocomotion/drake/blob/da4a09773e96859b322b90713c5e0790b3ce2ba9/drake/systems/controllers/inverse_dynamics.cc#L12)):*
> ```cpp
>                                     bool pure_gravity_compensation)
>     : tree_(tree),
>       pure_gravity_compensation{pure_gravity_compensation},
> ```

BTW, use `()` here rather than `{}` for consistency with the other initializers here.

---

*[drake/systems/controllers/inverse_dynamics.cc, line 71 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5339#-Ke63C53_kDaBdwzoYxu:-Ke63C53_kDaBdwzoYxv:b-j4fbf2) ([raw file](https://github.com/robotlocomotion/drake/blob/da4a09773e96859b322b90713c5e0790b3ce2ba9/drake/systems/controllers/inverse_dynamics.cc#L71)):*
> ```cpp
> template class InverseDynamics<double>;
> // TODO(siyuan): having linking error with this line 56
> // template class InverseDynamics<AutoDiffXd>;
> ```

Can you fix this now? I would have expected this to work.

---

*[drake/systems/controllers/inverse_dynamics.h, line 18 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5339#-Ke5vsb7VffvWDylVH1I:-Ke5vsb7VffvWDylVH1J:b7xegmr) ([raw file](https://github.com/robotlocomotion/drake/blob/da4a09773e96859b322b90713c5e0790b3ce2ba9/drake/systems/controllers/inverse_dynamics.h#L18)):*
> ```objc
>  * coriolis and centrifugal and gravity terms, `q` is the generalized position,
>  * `v` is the generalized velocity and `vd_d` is the desired generalized
>  * acceleration.
> ```

This needs more class documentation to explain what are the inputs and outputs, that there is a special mode for gravity compensation, and what equations are solved in gravity comp mode.

Should also mention InverseDynamicsController and when one should use this vs. that. Are these systems related?

---

*[drake/systems/controllers/inverse_dynamics_controller.cc, line 32 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5339#-Ke60zdZnOORBjQ0jgiU:-Ke60zdZnOORBjQ0jgiV:b-6d12vv) ([raw file](https://github.com/robotlocomotion/drake/blob/da4a09773e96859b322b90713c5e0790b3ce2ba9/drake/systems/controllers/inverse_dynamics_controller.cc#L32)):*
> ```cpp
> 
>   // Adds a PID.
>   auto pid = builder.template AddSystem<PidController<T>>(kp, ki, kd);
> ```

BTW, it would be very helpful to have a diagram here, however crude. It is hard to tell what your intention was by looking at the assortment of Connect() calls, and there is no way to see that these calls reflect that intention accurately.

---

*[drake/systems/controllers/inverse_dynamics_controller.h, line 15 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5339#-Ke5ztxFdYHSVhrmvAX-:-Ke5ztxFdYHSVhrmvAX0:b-3exe92) ([raw file](https://github.com/robotlocomotion/drake/blob/da4a09773e96859b322b90713c5e0790b3ce2ba9/drake/systems/controllers/inverse_dynamics_controller.h#L15)):*
> ```objc
>  * A state feedback controller whose output is:
>  * `torque = inverse_dynamics(q, v, vd_d)`,
>  * where `vd_d = kp * (q* - q) + kd * (v* - v) + ki * int(q* - q) + vd*`.
> ```

BTW, this is a bit awkward with `*` doing double duty as "reference" and "multiply". Consider leaving out the multiply operators or using some other symbol for reference (prime? bar as in q̅?). (Have you tried [unicodeit](http://www.unicodeit.net/) -- LaTeX in, Unicode out?)

---

*[drake/systems/controllers/inverse_dynamics_controller.h, line 21 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5339#-Ke60PEt5P_w4pox9WtQ:-Ke60PEt5P_w4pox9WtR:b-la8nng) ([raw file](https://github.com/robotlocomotion/drake/blob/da4a09773e96859b322b90713c5e0790b3ce2ba9/drake/systems/controllers/inverse_dynamics_controller.h#L21)):*
> ```objc
>  * Note that this class assumes the robot is fully actuated, its position
>  * and velocity have the same dimension, it does not have a floating base,
>  * and it does not contain any closed kinematic loops.
> ```

BTW, should say what happens if any of these conditions are violated.

---

*[drake/systems/controllers/inverse_dynamics_controller.h, line 24 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5339#-Ke61pfSSHmLHL2dn0Y7:-Ke61pfSSHmLHL2dn0Y8:b-pcr62x) ([raw file](https://github.com/robotlocomotion/drake/blob/da4a09773e96859b322b90713c5e0790b3ce2ba9/drake/systems/controllers/inverse_dynamics_controller.h#L24)):*
> ```objc
>  */
> template <typename T>
> class InverseDynamicsController : public ModelBasedController<T> {
> ```

Needs more class documentation explaining what are the inputs and outputs of this System and that it has an option to use zero for vd_d rather than read it from an input.

---

*[drake/systems/controllers/inverse_dynamics_controller.h, line 36 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5339#-Ke5wb5xELjA-3L69Bos:-Ke5wb5xELjA-3L69Bot:bglx74v) ([raw file](https://github.com/robotlocomotion/drake/blob/da4a09773e96859b322b90713c5e0790b3ce2ba9/drake/systems/controllers/inverse_dynamics_controller.h#L36)):*
> ```objc
>    * @param kp Position gain
>    * @param kp Velocity gain
>    * @param kp Integral gain
> ```

kp kp kp (here and below)

---

*[drake/systems/controllers/inverse_dynamics_controller.h, line 44 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5339#-Ke5yDciq47H9mkH-rvk:-Ke5yDciq47H9mkH-rvl:b-xz4bck) ([raw file](https://github.com/robotlocomotion/drake/blob/da4a09773e96859b322b90713c5e0790b3ce2ba9/drake/systems/controllers/inverse_dynamics_controller.h#L44)):*
> ```objc
>       std::shared_ptr<RigidBodyFrame<double>> world_offset,
>       const VectorX<T>& kp, const VectorX<T>& ki, const VectorX<T>& kd,
>       bool no_reference_acceleration);
> ```

BTW, I found this negative argument confusing (set no_blah to false to get blah). Consider calling this `use_reference_acceleration` and changing its sense.

---

*[drake/systems/controllers/test/inverse_dynamics_test.cc, line 165 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5339#-Ke64W8P4bn3nLocfelK:-Ke64W8P4bn3nLocfelL:bmcb2rn) ([raw file](https://github.com/robotlocomotion/drake/blob/da4a09773e96859b322b90713c5e0790b3ce2ba9/drake/systems/controllers/test/inverse_dynamics_test.cc#L165)):*
> ```cpp
>       tree.get());
>   EXPECT_DEATH(Init(std::move(tree), true), ".*");
> }
> ```

I'm not seeing unit tests for InverseDynamicsController?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5339#-:-Ke63U5Hei3EpnY64xjy:bjjtury)*
<!-- Sent from Reviewable.io -->

siyuanfeng-tri commented:
all capabilities of the gravity compensator are preserved with this pr. so if the user really wants to do just gravity compensation they still can.
i replaced the old one with this because i can't do inverse dynamics under the name gravity compensation. 
it's just much easier to specify gains on acceleration and have inverse dynamics scale those properly wrt different link inertias as opposed to specify the gains directly on torque. 

---

Review status: all files reviewed at latest revision, 12 unresolved discussions.

---

*[drake/examples/kuka_iiwa_arm/iiwa_wsg_simulation.cc, line 164 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5339#-Ke5xg7FUgMHTK7Nw5VK:-Ke67ZjV_s6291HSqN7T:bsn7drz) ([raw file](https://github.com/robotlocomotion/drake/blob/da4a09773e96859b322b90713c5e0790b3ce2ba9/drake/examples/kuka_iiwa_arm/iiwa_wsg_simulation.cc#L164)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW, "inertia gains"? Did you mean "integral gains"?
</blockquote></details>

should be integral 

---

*[drake/examples/kuka_iiwa_arm/controlled_kuka/kuka_demo_plant_builder.h, line 74 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5339#-Ke5sSmK7FqtCyF2FnEK:-Ke67ekuH1S3HVMtFhjL:b-r4qy55) ([raw file](https://github.com/robotlocomotion/drake/blob/da4a09773e96859b322b90713c5e0790b3ce2ba9/drake/examples/kuka_iiwa_arm/controlled_kuka/kuka_demo_plant_builder.h#L74)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW, this comment seems to lack context now. Previously there was a discussion about the desirability of derivatives and why they aren't used here but now that is gone. Please explain (resurrect the previous comment?).
</blockquote></details>

the previous message is not quite correct and can cause confusion. i want to avoid that. 
the issue basically boils down to: on the real robot, we can't command velocity or accelerations to the robot, but that's just an particular API restriction. the previous comments confused this with "the controller doesn't command velocity at all". i am 100% sure that the robot controller internally is generating velocity and acceleration commands from some combination of finite differencing and smoothing. 
since we are building a simulation for the closed loop system (the robot + it's internal controller), we ought to be able to generate those velocity commands, one way to do that is have the derivatives from the position trajectory. 

---

*[drake/systems/controllers/inverse_dynamics.cc, line 71 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5339#-Ke63C53_kDaBdwzoYxu:-Ke69UaGzu_as-UMZV8v:b76xans) ([raw file](https://github.com/robotlocomotion/drake/blob/da4a09773e96859b322b90713c5e0790b3ce2ba9/drake/systems/controllers/inverse_dynamics.cc#L71)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

Can you fix this now? I would have expected this to work.
</blockquote></details>

i have some linking trouble with KinematicsCache<T> cache = tree_.doKinematics(x.head(q_dim_), x.tail(v_dim_)); i suspect some explicit instantiation is missing, but it's very hard to track down. i can keep trying. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5339#-:-Ke663ggaljSMQnf9c8s:b-izp5bm)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
Platform review: a few minor BTWs below to take care of, then <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>.

---

Reviewed 1 of 5 files at r3, 12 of 12 files at r4.
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---

*[drake/examples/kuka_iiwa_arm/iiwa_wsg_simulation.cc, line 156 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/5339#-Ke7UuPnYcdng-zSnet3:-Ke7UuPnYcdng-zSnet4:b9cupmh) ([raw file](https://github.com/robotlocomotion/drake/blob/786c2d70455ac14a4367ce1583e13b57a6b0ee89/drake/examples/kuka_iiwa_arm/iiwa_wsg_simulation.cc#L156)):*
> ```cpp
>     VectorX<double> iiwa_kp, iiwa_kd, iiwa_ki;
>     SetPositionControlledIiwaGains(&iiwa_kp, &iiwa_ki, &iiwa_kd);
>     // Uses intergral gains to deal with the added mass from the grasped object.
> ```

BTW, intergral -> integral

---

*[drake/systems/controllers/inverse_dynamics.cc, line 71 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5339#-Ke63C53_kDaBdwzoYxu:-Ke7XxcEJFYfM7QJJVUy:b-mp9y3z) ([raw file](https://github.com/robotlocomotion/drake/blob/da4a09773e96859b322b90713c5e0790b3ce2ba9/drake/systems/controllers/inverse_dynamics.cc#L71)):*
<details><summary><i>Previously, siyuanfeng-tri wrote…</i></summary><blockquote>

i have some linking trouble with KinematicsCache<T> cache = tree_.doKinematics(x.head(q_dim_), x.tail(v_dim_)); i suspect some explicit instantiation is missing, but it's very hard to track down. i can keep trying. 
</blockquote></details>

BTW, I would like to see this fixed now, but if it seems too hard I would be satisfied with an issue filed and referenced in the TODO.

---

*[drake/systems/controllers/inverse_dynamics.h, line 22 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/5339#-Ke7Vw4TruwGc65KHU5L:-Ke7Vw4TruwGc65KHU5M:bitzkpx) ([raw file](https://github.com/robotlocomotion/drake/blob/786c2d70455ac14a4367ce1583e13b57a6b0ee89/drake/systems/controllers/inverse_dynamics.h#L22)):*
> ```objc
>  * input port for the state `(q, v)` and an BasicVector output port for the
>  * computed torque. There is an additional BasicVector input port for desired
>  * acceleration when configured to be not in pure gravity compensation mode.
> ```

BTW, still needs a "see also" regarding the InverseDynamicsController -- how are these classes related? Which one should an API user use under what circumstances?

---

*[drake/systems/controllers/inverse_dynamics_controller.cc, line 40 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/5339#-Ke7WH_hrlmSX98RQRZK:-Ke7WH_hrlmSX98RQRZL:b-plslc1) ([raw file](https://github.com/robotlocomotion/drake/blob/786c2d70455ac14a4367ce1583e13b57a6b0ee89/drake/systems/controllers/inverse_dynamics_controller.cc#L40)):*
> ```cpp
>              |                   | inverse dynamics | ---> torque
>              ------------------> |                  |
> ```

BTW, great -- thanks!

---

*[drake/systems/controllers/inverse_dynamics_controller.h, line 22 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/5339#-Ke7WfEx3WN7NCKrrKA_:-Ke7WfEx3WN7NCKrrKAa:b-gva3j8) ([raw file](https://github.com/robotlocomotion/drake/blob/786c2d70455ac14a4367ce1583e13b57a6b0ee89/drake/systems/controllers/inverse_dynamics_controller.h#L22)):*
> ```objc
>  * the generalized acceleration. `*` indicates reference values.
>  *
>  * This controller always a BasicVector input port for estimated robot state
> ```

BTW, missing "has"

---

*[drake/systems/controllers/inverse_dynamics_controller.h, line 80 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/5339#-Ke7X0Hmb-HCfc2slJb6:-Ke7X0Hmb-HCfc2slJb7:blum9et) ([raw file](https://github.com/robotlocomotion/drake/blob/786c2d70455ac14a4367ce1583e13b57a6b0ee89/drake/systems/controllers/inverse_dynamics_controller.h#L80)):*
> ```objc
>    * @param ki Integral gain
>    * @param kd Velocity gain
>    * @param has_reference_acceleration If true, There is an extra BasicVector
> ```

BTW, There -> there

---

*[drake/systems/controllers/inverse_dynamics_controller.h, line 108 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/5339#-Ke7X8gTYV8hfbSW4VnQ:-Ke7X8gTYV8hfbSW4VnR:b-w09nn2) ([raw file](https://github.com/robotlocomotion/drake/blob/786c2d70455ac14a4367ce1583e13b57a6b0ee89/drake/systems/controllers/inverse_dynamics_controller.h#L108)):*
> ```objc
> 
>   PidController<T>* pid_{nullptr};
>   const bool has_reference_acceleration_{false};
> ```

BTW, did you mean to change the default? It now says the default has no reference acceleration but before the default was yes.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5339#-:-Ke7YLmPFtHVauYCMIsv:b-vesg78)*
<!-- Sent from Reviewable.io -->

siyuanfeng-tri commented:
Review status: 22 of 26 files reviewed at latest revision, 16 unresolved discussions.

---

*[drake/examples/kuka_iiwa_arm/iiwa_wsg_simulation.cc, line 164 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5339#-Ke5xg7FUgMHTK7Nw5VK:-Ke6LD00Sc53kBq0XG1l:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/da4a09773e96859b322b90713c5e0790b3ce2ba9/drake/examples/kuka_iiwa_arm/iiwa_wsg_simulation.cc#L164)):*
<details><summary><i>Previously, siyuanfeng-tri wrote…</i></summary><blockquote>

should be integral 
</blockquote></details>

Done.

---

*[drake/examples/kuka_iiwa_arm/iiwa_wsg_simulation.cc, line 156 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/5339#-Ke7UuPnYcdng-zSnet3:-Ke96TtFoNB3eZOfKK0I:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/786c2d70455ac14a4367ce1583e13b57a6b0ee89/drake/examples/kuka_iiwa_arm/iiwa_wsg_simulation.cc#L156)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW, intergral -> integral
</blockquote></details>

Done.

---

*[drake/systems/controllers/inverse_dynamics.cc, line 12 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5339#-Ke5trSzycDa_o2zxlub:-Ke6LL0zhmgPKaeTbK8j:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/da4a09773e96859b322b90713c5e0790b3ce2ba9/drake/systems/controllers/inverse_dynamics.cc#L12)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW, use `()` here rather than `{}` for consistency with the other initializers here.
</blockquote></details>

Done.

---

*[drake/systems/controllers/inverse_dynamics.cc, line 71 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5339#-Ke63C53_kDaBdwzoYxu:-Ke9EpVkx62_aLfbJApe:b4re5y2) ([raw file](https://github.com/robotlocomotion/drake/blob/da4a09773e96859b322b90713c5e0790b3ce2ba9/drake/systems/controllers/inverse_dynamics.cc#L71)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW, I would like to see this fixed now, but if it seems too hard I would be satisfied with an issue filed and referenced in the TODO.
</blockquote></details>

hmm, i rebased onto last night's master, and it seems to work now. hmmm.

---

*[drake/systems/controllers/inverse_dynamics.h, line 18 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5339#-Ke5vsb7VffvWDylVH1I:-Ke6TFmKNfIVFZZniE0T:b-o7cr3j) ([raw file](https://github.com/robotlocomotion/drake/blob/da4a09773e96859b322b90713c5e0790b3ce2ba9/drake/systems/controllers/inverse_dynamics.h#L18)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

This needs more class documentation to explain what are the inputs and outputs, that there is a special mode for gravity compensation, and what equations are solved in gravity comp mode.

Should also mention InverseDynamicsController and when one should use this vs. that. Are these systems related?
</blockquote></details>

Done. added more comments in inverse dynamics controller.h

---

*[drake/systems/controllers/inverse_dynamics.h, line 22 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/5339#-Ke7Vw4TruwGc65KHU5L:-Ke97oXIpMFou4_NqK5x:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/786c2d70455ac14a4367ce1583e13b57a6b0ee89/drake/systems/controllers/inverse_dynamics.h#L22)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW, still needs a "see also" regarding the InverseDynamicsController -- how are these classes related? Which one should an API user use under what circumstances?
</blockquote></details>

Done.

---

*[drake/systems/controllers/inverse_dynamics_controller.cc, line 32 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5339#-Ke60zdZnOORBjQ0jgiU:-Ke6WJCpPsqfe56pIVDe:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/da4a09773e96859b322b90713c5e0790b3ce2ba9/drake/systems/controllers/inverse_dynamics_controller.cc#L32)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW, it would be very helpful to have a diagram here, however crude. It is hard to tell what your intention was by looking at the assortment of Connect() calls, and there is no way to see that these calls reflect that intention accurately.
</blockquote></details>

Done.

---

*[drake/systems/controllers/inverse_dynamics_controller.cc, line 40 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/5339#-Ke7WH_hrlmSX98RQRZK:-Ke97pM0bt1F3RD29_5F:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/786c2d70455ac14a4367ce1583e13b57a6b0ee89/drake/systems/controllers/inverse_dynamics_controller.cc#L40)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW, great -- thanks!
</blockquote></details>

Done.

---

*[drake/systems/controllers/inverse_dynamics_controller.h, line 15 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5339#-Ke5ztxFdYHSVhrmvAX-:-Ke6P6dvA_wGRmybeKq0:bqi2fg4) ([raw file](https://github.com/robotlocomotion/drake/blob/da4a09773e96859b322b90713c5e0790b3ce2ba9/drake/systems/controllers/inverse_dynamics_controller.h#L15)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW, this is a bit awkward with `*` doing double duty as "reference" and "multiply". Consider leaving out the multiply operators or using some other symbol for reference (prime? bar as in q̅?). (Have you tried [unicodeit](http://www.unicodeit.net/) -- LaTeX in, Unicode out?)
</blockquote></details>

i dropped multiplication. 

---

*[drake/systems/controllers/inverse_dynamics_controller.h, line 21 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5339#-Ke60PEt5P_w4pox9WtQ:-Ke6WL4Ot-kgVt7mJHLr:b-eerp9o) ([raw file](https://github.com/robotlocomotion/drake/blob/da4a09773e96859b322b90713c5e0790b3ce2ba9/drake/systems/controllers/inverse_dynamics_controller.h#L21)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW, should say what happens if any of these conditions are violated.
</blockquote></details>

added more comments.

---

*[drake/systems/controllers/inverse_dynamics_controller.h, line 24 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5339#-Ke61pfSSHmLHL2dn0Y7:-Ke6WN6oDsK62HhUe0o6:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/da4a09773e96859b322b90713c5e0790b3ce2ba9/drake/systems/controllers/inverse_dynamics_controller.h#L24)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

Needs more class documentation explaining what are the inputs and outputs of this System and that it has an option to use zero for vd_d rather than read it from an input.
</blockquote></details>

Done.

---

*[drake/systems/controllers/inverse_dynamics_controller.h, line 36 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5339#-Ke5wb5xELjA-3L69Bos:-Ke6RQNnR5TkxiHwScVX:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/da4a09773e96859b322b90713c5e0790b3ce2ba9/drake/systems/controllers/inverse_dynamics_controller.h#L36)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

kp kp kp (here and below)
</blockquote></details>

Done.

---

*[drake/systems/controllers/inverse_dynamics_controller.h, line 44 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5339#-Ke5yDciq47H9mkH-rvk:-Ke6TDFk6dLEvry5VKwm:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/da4a09773e96859b322b90713c5e0790b3ce2ba9/drake/systems/controllers/inverse_dynamics_controller.h#L44)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW, I found this negative argument confusing (set no_blah to false to get blah). Consider calling this `use_reference_acceleration` and changing its sense.
</blockquote></details>

Done.

---

*[drake/systems/controllers/inverse_dynamics_controller.h, line 22 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/5339#-Ke7WfEx3WN7NCKrrKA_:-Ke97woBT34ZUJZPd_h_:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/786c2d70455ac14a4367ce1583e13b57a6b0ee89/drake/systems/controllers/inverse_dynamics_controller.h#L22)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW, missing "has"
</blockquote></details>

Done.

---

*[drake/systems/controllers/inverse_dynamics_controller.h, line 80 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/5339#-Ke7X0Hmb-HCfc2slJb6:-Ke982_MzVs068tiGQ_G:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/786c2d70455ac14a4367ce1583e13b57a6b0ee89/drake/systems/controllers/inverse_dynamics_controller.h#L80)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW, There -> there
</blockquote></details>

Done.

---

*[drake/systems/controllers/inverse_dynamics_controller.h, line 108 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/5339#-Ke7X8gTYV8hfbSW4VnQ:-Ke98872ikUruXsR7wtj:bd7kmf4) ([raw file](https://github.com/robotlocomotion/drake/blob/786c2d70455ac14a4367ce1583e13b57a6b0ee89/drake/systems/controllers/inverse_dynamics_controller.h#L108)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW, did you mean to change the default? It now says the default has no reference acceleration but before the default was yes.
</blockquote></details>

I think this flag needs to be pass to the constructor no matter what. So this default value shouldn't matter. I merely set it to have it initialized at compile time. 

---

*[drake/systems/controllers/test/inverse_dynamics_test.cc, line 165 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5339#-Ke64W8P4bn3nLocfelK:-Ke6WTKk0Rz9iwAdlJyQ:b-jfiq7e) ([raw file](https://github.com/robotlocomotion/drake/blob/da4a09773e96859b322b90713c5e0790b3ce2ba9/drake/systems/controllers/test/inverse_dynamics_test.cc#L165)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

I'm not seeing unit tests for InverseDynamicsController?
</blockquote></details>

working on a simple one. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5339)*
<!-- Sent from Reviewable.io -->

siyuanfeng-tri commented:
*[drake/systems/controllers/inverse_dynamics.cc, line 71 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5339#-Ke63C53_kDaBdwzoYxu:-Ke9JSd3rVGIHcP8o-6W:b8bnj0a) ([raw file](https://github.com/robotlocomotion/drake/blob/da4a09773e96859b322b90713c5e0790b3ce2ba9/drake/systems/controllers/inverse_dynamics.cc#L71)):*
<details><summary><i>Previously, siyuanfeng-tri wrote…</i></summary><blockquote>

hmm, i rebased onto last night's master, and it seems to work now. hmmm.
</blockquote></details>

no wait, it still has the linking error with cmake, but not with bazel. hmm. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5339)*
<!-- Sent from Reviewable.io -->

siyuanfeng-tri commented:
@sherm1 i wasn't really sure what i was doing with the explicit instantiation stuff in RBT, would be nice if you can double check that. 
also for some reason, i can't directly do tree.doKinematics(q, v), which also produced a linking error. instead i had to do cache = tree.CreateKinematicsCache(), cache.init(q,v), tree.doKinematics(cache). 

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5339#-:-Ke9ScAwfUfanl3r-CiW:b-y2rmec)*
<!-- Sent from Reviewable.io -->

siyuanfeng-tri commented:
also, i didn't have any trouble in bazel land, all these weird stuff happened with cmake / ninja. 

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5339#-:-Ke9TAbabgLdj3wHCqiH:bmczw9i)*
<!-- Sent from Reviewable.io -->

siyuanfeng-tri commented:
@drake-jenkins-bot linux-xenial-clang-ninja-experimental-minimal please

sherm1 commented:
> i wasn't really sure what i was doing with the explicit instantiation stuff in RBT, would be nice if you can double check that

I suggest talking to @amcastro-tri -- I believe he knows his way around that code well.

This PR is good to go I think, except for the unit test you said you were going to add. Also if you can't get the AutoDiff stuff to work please file an issue and reference it in a TODO here. 

---

Reviewed 5 of 5 files at r5.
Review status: :shipit: all files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5339#-:-KeAh5hHUWU3snDQz8sg:b-7kx845)*
<!-- Sent from Reviewable.io -->

siyuanfeng-tri commented:
i did add a simple there for the controller there. https://reviewable.io/reviews/robotlocomotion/drake/5339/drake/systems/controllers/test/inverse_dynamics_controller_test.cc
short story i got the AutoDiff version to compile with cmake, but i have not convinced that's correct.. i briefly talked with @jwnimmer-tri, and we sort of agreed that whatever makes it compiled is probably fine for now. 

thanks again sherm!

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5339#-:-KeAicIQSNjOXkhX--Io:byqva81)*
<!-- Sent from Reviewable.io -->
